torch/csrc/jit/autodiff.cpp

```diff
@@ -202,6 +229,45 @@ static std::shared_ptr<Graph> splitOffStage(
   return graph_clone;
 }
 
+bool isZero(Value * v) {
+  auto n = v->node();
+  return n->kind() == kConstant &&
+    n->hasAttribute(kis_zero) &&
+    n->i(kis_zero);
+}
+
+// In the case where an input is routed to an output
+// return the (possibly undefined) input rather than
+// the value guarded by replaceIfUndef
+// this ensures that we do not produce a 0 tensor
+// when the autograd would produce None
+// graph(a) {
+//   b = replaceIfUndef(a,0);
+//   c = b + b
+//   return c, b; // will replace 'b' with 'a'
+// }
+// Also replace any known-to-be-zero outputs with Undef
+// for the same reason
```

Comment:
I am not sure if that is true in practice then. If I remove the known-to-be-zero check, then I get mismatches where I return zeroes but the autograd returns None.  I don't think we produce outputs in `df` for inputs that do not require gradients, so that is not the pathway being taken here. Any ideas what is going on?

Reply:
A follow up: using Undefined to sometimes-but-not-always represent a 0 is very fragile and impacts the entire codebase. We really should have a concept of a known-to-be-zero zero-dim tensor (i.e. a zero-literal-tensor) that is a subtype of all other tensor types but can be tested for in ops that can optimize for it.